### PR TITLE
Adding Switch cluster - Issue #91

### LINKED ISSUE #91: Add Switch Cluster
I'm working this one.

### PR DESCRIPTION
**Pull request by JimBuzbee**
Initial implementation of Switch Cluster

fixes #91

### CODE REVIEW

File: packages/matter.js/src/cluster/SwitchCluster.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        /** Indicates a switch that does not maintain its position after being pressed (or turned)  */
+        momentarySwitch: BitFlag(1),
+
+        /** Indicates a momentary switch that can distinguish and report release events. */
+        momentarySwitchRelease: BitFlag(2),
+
+        /** Indicates a momentary switch that can distinguish and report long presses from short presses. */
+        momentarySwitchLongPress: BitFlag(3),
+
+        /** Idicates a momentary switch that can distinguish and report double press
```
Comment (Apollon77) [inline packages/matter.js/src/cluster/SwitchCluster.ts:37]:
```suggestion
        /** 
         * Idicates a momentary switch that can distinguish and report double press 
```

File: packages/matter.js/src/cluster/SwitchCluster.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    /** @see {@link MatterApplicationClusterSpecificationV1_0} § 1.11.5 */
+    attributes: {
+        /** Indicates the maximum number of positions the switch has with a minimum of 2 */
+        numberOfPositions: Attribute(0, TlvUInt8.bound({ min: 2 }), { default: 2 }),
+
+        /** Indicates the position of the switch with a valid range of zero to NumberOfPositions-1 */
+        currentPosition: Attribute(1, TlvUInt8, { default: 0 }),
+
+        /** Indicates how many consecutive presses can be detected/reported by a multi-press switch */
+        multiPressMax: OptionalAttribute(2, TlvUInt8.bound({ min: 2 }), { default: 2 }),
```
Comment (Apollon77) [inline packages/matter.js/src/cluster/SwitchCluster.ts:52]: Add TODO please with info that mandatory is MSM feature is set ... same please for all evenet below about their constrains

File: packages/matter.js/src/cluster/SwitchCluster.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        LongPress: OptionalEvent(2, EventPriority.Info, { newPosition: TlvField(0, TlvUInt8) }),
+
+        /** Generated when the momentary switch has been released. */
+        ShortRelease: OptionalEvent(3, EventPriority.Info, { previousPosition: TlvField(0, TlvUInt8) }),
+
+        /** Gnerated when the momentary switch has been released. */
+        LongRelease: OptionalEvent(4, EventPriority.Info, { previousPosition: TlvField(0, TlvUInt8) }),
+
+        /** Indicates how many times the switch has been pressed in a multi-press sequence during that sequence */
+        MultiPressOngoing: OptionalEvent(5, EventPriority.Info, { newPosition: TlvField(0, TlvUInt8), currentNumberOfPressesCounted: TlvField(1, TlvUInt8) }),
```
Comment (Apollon77) [inline packages/matter.js/src/cluster/SwitchCluster.ts:73]: currentNumberOfPressesCounted has min 2 as bound

File: packages/matter.js/src/cluster/SwitchCluster.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+        /** Indicates how many times the switch has been pressed after it has been detected that the sequence has ended. */
+        MultiPressComplete: OptionalEvent(6, EventPriority.Info, { previousPosition: TlvField(0, TlvUInt8), totalNumberOfPressesCounted: TlvField(1, TlvUInt8) }),
```
Comment (Apollon77) [inline packages/matter.js/src/cluster/SwitchCluster.ts:76]: totalNumberOfPressesCounted has min 1 as bound

**Review by Apollon77 — CHANGES_REQUESTED** [review]
just minor stuff, thank you!

File: packages/matter.js/src/cluster/SwitchCluster.ts (hunk 2 above)
Comment (JimBuzbee) [reply packages/matter.js/src/cluster/SwitchCluster.ts]: Added a bunch of TODOs - kept them on the same line as needs work, but some lines then get very long. Maybe there is a better way to format?

File: packages/matter.js/src/cluster/SwitchCluster.ts (hunk 4 above)
Comment (Apollon77) [reply packages/matter.js/src/cluster/SwitchCluster.ts]: why you used min in quotes?

File: packages/matter.js/src/cluster/SwitchCluster.ts (hunk 3 above)
Comment (Apollon77) [reply packages/matter.js/src/cluster/SwitchCluster.ts]: why you used min in quotes?

File: packages/matter.js/src/cluster/SwitchCluster.ts (hunk 2 above)
Comment (Apollon77) [reply packages/matter.js/src/cluster/SwitchCluster.ts]: Yes I see ... alternative is to move comment to abobe the /** ...*/ line

**Review by Apollon77 — COMMENTED** [review]
Commented ... pleae fix the "min" thingy in any cae ... the TODO I leave up to you.

I would leave the PR maybe unmerged ... @mfucci is working on better handling of feature specific attributes/commands/events in #87 ... maybe we could then bring it in directly "correct"

File: packages/matter.js/src/cluster/SwitchCluster.ts (hunk 3 above)
Comment (JimBuzbee) [reply packages/matter.js/src/cluster/SwitchCluster.ts]: I don't know :-)  Will fix both!

**Comment by Apollon77** [comment]
Thank you, basically approved. As said I would like to leave unmerged for now to ait for @mfucci s PR ... Then we cn direct,y rework it correctly

**Comment by Apollon77** [comment]
@JimBuzbee please also add it to https://github.com/project-chip/matter.js/blob/main/packages/matter.js/src/cluster/ClusterHelper.ts#L35

**Pull request closed without merging**